src/main/kotlin/org/elm/lang/core/types/TypeReplacement.kt

```diff
@@ -58,7 +59,29 @@ class TypeReplacement(
          */
         fun flexify(ty: Ty): Ty {
             if (ty.allVars(true).none()) return ty
-            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList()).replace(ty)
+            return TypeReplacement(emptyMap(), freshen = false, varsToRemainRigid = emptyList(), freeze = false).replace(ty)
+        }
+
+        /**
+         * Freeze any fields references present in records in a [ty].
+         *
+         * Use this to prevent cached values from being altered.
+         */
+        fun freeze(ty: Ty): Ty {
```

Comment:
I don't know if it great that we can reuse type replacement or so many things, or problematic that we do.